rest-client/src/main/java/org/sagebionetworks/bridge/rest/UserSessionInfoProvider.java

```diff
@@ -85,28 +88,48 @@ synchronized void reauthenticate() throws IOException {
                 UserSessionInfo newSession = authenticationApi.reauthenticate(request).execute().body();
                 setSession(newSession);
             } catch(ConsentRequiredException e) {
+                // successful authentication
                 setSession(e.getSession());
-            } catch(Exception e) {
+            } catch(BridgeSDKException reauthException) {
                 if (LOG.isDebugEnabled()) {
-                    LOG.debug("Reauth failed, signing in again", e);
+                    LOG.debug("Reauth failed, signing in again", reauthException);
+                }
+
+                boolean signInAttempted;
+                try {
+                     signInAttempted = signIn();
+                } catch (BridgeSDKException signInException) {
+                    try {
+                        signInException.initCause(reauthException);
+                    } catch (IllegalStateException t) {
+                        // should not happen, since cause should not have been set before
+                        LOG.debug("Unable to set cause on: ", signInException);
+                    }
+                    throw signInException;
+                }
+
+                // if we never try to sign in, propagate exception from reauth
+                if (!signInAttempted) {
+                    throw reauthException;
```

Comment:
At a higher level than this, do we want to sign in automatically for people when we can't reauthenticate? Don't we want people to go through the sign in process again, so we never have to store their credentials?

Reply:
That is a good point, and with reauth token, we can now do that. I'll file a JIRA but I'd prefer not to introduce this to the apps going out soon.